libyul/SideEffects.cpp

```diff
@@ -0,0 +1,178 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Specific AST walkers that collect semantical facts.
+ */
+
+#include <libyul/optimiser/Semantics.h>
+
+#include <libyul/Exceptions.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+#include <libyul/backends/evm/EVMDialect.h>
+
+#include <libevmasm/SemanticInformation.h>
+
+#include <libdevcore/CommonData.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+SideEffectsCollector::SideEffectsCollector(Dialect const& _dialect, Expression const& _expression):
+	SideEffectsCollector(_dialect)
+{
+	visit(_expression);
+}
+
+SideEffectsCollector::SideEffectsCollector(Dialect const& _dialect, Statement const& _statement):
+	SideEffectsCollector(_dialect)
+{
+	visit(_statement);
+}
+
+SideEffectsCollector::SideEffectsCollector(Dialect const& _dialect, Block const& _ast):
+	SideEffectsCollector(_dialect)
+{
+	operator()(_ast);
+}
+
+void SideEffectsCollector::operator()(FunctionalInstruction const& _instr)
+{
+	ASTWalker::operator()(_instr);
+
+	if (!eth::SemanticInformation::movable(_instr.instruction))
+		m_sideEffects.movable = false;
+	if (!eth::SemanticInformation::sideEffectFree(_instr.instruction))
+		m_sideEffectFree = false;
```

Comment:
Now I'm confused - am I looking at the right header for this? Aren't these members removed there and it should access indirectly via ``m_sideEffects`` now as above for ``movable``? If so why does this PR even build? I'm stopping my review here for now, until this is clarified..